### PR TITLE
Add post author avatar block

### LINKED ISSUE #31962: Post Author Block: Inconsistent Front/Back End Avatar Markup and Classes
The `` tag for the Post Author block avatar outputs the following on the front end:

```
<img alt="" src="#" class="avatar avatar-48 photo" height="48" width="48" loading="lazy">
```

In the editor, the HTML looks like:

```
<img width="48" src="#" alt="Justin Tadlock">
```

The editor should at least have the same `.avatar` class.

Also, the `alt` text should match.  Is there a reason it's empty on the front end?

### PR DESCRIPTION
**Pull request by carolinan**
## Description

Adds a new block for the post author avatar.
Partial for https://github.com/WordPress/gutenberg/issues/35596
Also closes https://github.com/WordPress/gutenberg/issues/31962, https://github.com/WordPress/gutenberg/issues/35693

The goal is to support the following:

- Resizing
- Border
- Link to author archive
- Duotone
- Margin
- None/Left/center/right align
- Bonus: Display a placeholder while loading

_Compared to the comment avatar I opted out of padding and background color because I did not see how the background color would be used with this block._

The block has a wrapping figure element, an optional link, and an image element.

The wrapper is needed for two reasons:

- The margin needs to be placed outside the ResizableBox used for resizing in the editor, or it affects the size.
- The duotone can only be applied to an element inside a block, not the block itself. (I tried different selectors and `"__experimentalSelector"`, [but the documentation says that the parent selector is automatically included](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#color-__experimentalduotone) )

**Problems:**

- The duotone and the spacing styles need to be placed on the wrapper.
- The border styles need to be placed only on the image element, if it is placed on the wrapper it does not achieve the wanted effect.
- In the editor, only top and right margin works, because the wrapper does not have a width.
_Adding width: fit-content to the wrapper may solve this._ On the front, all four sides of the margin works, because it is added to the image element. This does lead to inconsistencies between the editor and front.

**Question:**
How do I remove the "has-border-color" CSS class from blockProps, which is placed on the wrapper?
I read through the documentation for _classnames_ and tried several things but was not able to remove it.

**Alternatives:**
The block could include a CSS file to force hide the border on the wrapper but that did not feel like the right approach either.

## Testing Instructions

1. Open a post or page in the editor
2. Insert a post author avatar block
3. Duplicate the block and try the different settings, including colors.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

File: docs/reference-guides/core-blocks.md [hunk 1]
```diff
@@ -159,7 +159,7 @@ Contains the block elements used to render a comment, like the title, date, auth
 -	**Name:** core/comment-template
 -	**Category:** design
 -	**Supports:** align, ~~html~~, ~~reusable~~
--	**Attributes:**
```
Comment (carolinan) [inline docs/reference-guides/core-blocks.md]: Not sure why these spacing changes were included :/

**Comment by cbravobernal** [comment]
> Question:
How do I remove the "has-border-color" CSS class from blockProps, which is placed on the wrapper?
I read through the documentation for classnames and tried several things but was not able to remove it.

Not sure if it is the best way to do it. But if you add `"__experimentalSkipSerialization": true,` on border props. They should be not applied when you call `useBlockProps();`.

That way you can get border props (both classname and styles) by:
  - `import { __experimentalGetBorderClassesAndStyles as getBorderClassesAndStyles } from '@wordpress/block-editor';`
  - `const borderProps = getBorderClassesAndStyles(attributes);`
  - Then apply those props wherever you need them.

You have a similar approach on `packages/block-library/build-module/button/edit.js`

We did something similar to the Comment Author Avatar, feel free to check it, it was with spacing attributes:

- [PHP part](https://github.com/WordPress/gutenberg/blob/b3e57fb20d9ebdb87e7644a58314a0191e66aeb5/packages/block-library/src/comment-author-avatar/index.php#L29-L47)
- [Skip Serialization on the block.json](https://github.com/WordPress/gutenberg/blob/b3e57fb20d9ebdb87e7644a58314a0191e66aeb5/packages/block-library/src/comment-author-avatar/block.json#L34)
- [Calling spacing attributes in edit.js](https://github.com/WordPress/gutenberg/blob/b3e57fb20d9ebdb87e7644a58314a0191e66aeb5/packages/block-library/src/comment-author-avatar/edit.js#L39)

**Comment by cbravobernal** [comment]
I've also added here a "post featured image block" like to show the Comment Avatar placeholder if we want to have it similar. Done in PR #38072 

Screenshot 2022-01-27 at 15 06 00

https://github.com/WordPress/gutenberg/pull/38072/commits/e5127d74ad084513b94a0a185e9514e540fcfc70

**Comment by carolinan** [comment]
@c4rl0sbr4v0 Thank you for the help! I think it is solved now. I added `__experimentalSkipSerialization` to the border in block.json and then found the solution for re-adding the border attributes in the code for the search block.

I tested using the placeholder SVG; but found that on my install it was never visible. I am not sure how to measure it, but perhaps the svg takes so long to load that the correct image is displayed just as fast.
So I switched back to showing the placeholder text only.

**Comment by carolinan** [comment]
I would like to avoid the two stylesheets, if there are better ways to solve the left and right margins that are not working in the editor without the CSS.

I feel like being able to adjust the spacing will be important for creating different layouts together with the author name and biography blocks.

**Comment by cbravobernal** [comment]
Great job! I will take this PR as a reference for the comments too 🚀 

Sorry, I don't completely understand the issue with SVG you mention. 🙇

> I tested using the placeholder SVG; but found that on my install it was never visible. I am not sure how to measure it, but perhaps the svg takes so long to load that the correct image is displayed just as fast. So I switched back to showing the placeholder text only.

We could make that square only with plain CSS if it is faster than a SVG. Also, we need to add this CSS to the SVG parent dom element  in order to be shown:
```css 
stroke: currentColor;
stroke-dasharray: 3;
```

> the correct image is displayed just as fast

Do you mean the real post author avatar?

**Comment by carolinan** [comment]
> > the correct image is displayed just as fast
> 
> Do you mean the real post author avatar?

Yes.

**Comment by cbravobernal** [comment]
> > > the correct image is displayed just as fast
> > 
> > 
> > Do you mean the real post author avatar?
> 
> Yes.

In that case, the main issue we have, is that if we go to the Site Editor -> Single Post. We cannot give styles to a text placeholder. (screenshot attached)

Screenshot 2022-01-28 at 12 08 09

@jameskoster [mentioned the idea of displaying a default avatar](https://github.com/WordPress/gutenberg/pull/38072#issuecomment-1024031679), I'm still checking if that would mean to do a request to the REST API to fetch that information. Will keep you posted here as well!

**Comment by carolinan** [comment]
The author avatar has a fallback image already if the user has not set up an avatar at gravatar.
The text that is displayed before the avatar loads does not need to be customizable because it displays for 1-2 seconds _(on my connection_). It only flashes.

**Comment by cbravobernal** [comment]
>  it displays for 1-2 seconds (on my connection). It only flashes.

At least, this is also happening on Comments Avatar Author on some template parts edition, and also inside any post or page 😅 

But, I recorded a screen video where when you edit the Single Post template on the Site Editor, it does not display an avatar, just the text placeholder.

https://user-images.githubusercontent.com/37012961/151554988-28de1b16-e289-4478-8b05-d1623ffc1f37.mov

**Comment by carolinan** [comment]
Would it be acceptable for the placeholder to keep the .4 opacity? It affects the border color option but only in the editor, so it doesn't match the front.

Or would it be better if the placeholder respected the border color option, but looked different from all the other placeholders that use .4 opacity?

**Comment by carolinan** [comment]
Left: Editor, various settings. Right: front.
Here the difference in opacity is visible.

For some reason, the front is displaying an avatar from an existing user (from the imported theme test data).
It is not the logged in user, and it is also not the author of the first post in the post list (I used the index template for the test and it did have a query loop below the avatars. _The avatars are not placed in the query loop._) 🤔 

avatar placeholders with different borders in the editor, and the same block on the front.

**Comment by cbravobernal** [comment]
> Or would it be better if the placeholder respected the border color option, but looked different from all the other placeholders that use .4 opacity?

I guess the best option is always to keep consistency between editor and frontend. So in this case would be better to remove that opacity. That will also happen to comments author's avatar.

**Comment by carolinan** [comment]
I am trying to troubleshoot why the avatar is displaying an actual user avatar on the front, for pages that are not single post type templates.

1. Install the [theme test data](https://github.com/WPTT/theme-test-data)
2. Open the index template in the site editor and add a post author avatar block without a container. See the placeholder in the editor.
3. View the front. Here, I am seeing the avatar for the user "themedemos" with the display name "Theme buster".

At this point, I added a `var_dump()` of the post id and author id inside the blocks index.php file.
I expected both these values to not be set since it is not a single post type.
The result is the post and author id of the "Template: Sticky" post from the theme test data.

4. Delete the query loop in the index. Save and view the front. The avatar, post id and author id are still the same.

**Comment by carolinan** [comment]
Perhaps its a non issue if the blocks will be limited to a parent post author block.

**Comment by jameskoster** [comment]
> Compared to the comment avatar I opted out of padding and background color because I did not see how the background color would be used with this block.

Shouldn't they be consistent? Padding and background enable you to create things like this:

Screenshot 2022-02-07 at 11 01 27

IE some spacing around the image _and_ a border.

**Comment by carolinan** [comment]
I was not able to do that because the padding is inside the resizer, but the border is inside.

**Comment by carolinan** [comment]
I was not able to get that to work with the resizer.

**Comment by carolinan** [comment]
Anyway I am hesitant to do any more work on this until there is a final decision about if this is going to be separate block or a variation of the post comment avatar block, and what the new name will be and so on.

**Comment by jasmussen** [comment]
For what it's worth, I think it's valuable to have a simple avatar block like this 👍 👍

**Comment by cbravobernal** [comment]
> Anyway I am hesitant to do any more work on this until there is a final decision about if this is going to be separate block or a variation of the post comment avatar block, and what the new name will be and so on.

@carolinan I will push at the end of the day a draft PR of a first iteration. I'm comparing and taking code from both Comment and Post, and will be happy to share the development, meaning you can commit without any hesitation, your experience is kindly appreciated 😄

**Comment by carolinan** [comment]
Closed in favor of https://github.com/WordPress/gutenberg/pull/38591

**Pull request closed without merging**